pkg/util/lsof/lsof_linux.go

```diff
@@ -0,0 +1,383 @@
+	5:  "FIN_WAIT2",
+	6:  "TIME_WAIT",
+	7:  "CLOSE",
+	8:  "CLOSE_WAIT",
+	9:  "LAST_ACK",
+	10: "LISTEN",
+	11: "CLOSING",
+	12: "NEW_SYN_RECV",
+	13: "BOUND_INACTIVE",
+}
+
+func stateStr(state uint64) string {
+	if s, ok := tcpStates[state]; ok {
+		return s
+	}
+	return fmt.Sprintf("UNKNOWN(%d)", state)
+}
+
+// readSocketInfo reads the socket information from /proc/<pid>/net/{tcp,tcp6,udp,udp6,unix}
+// returns a map of inode to socketInfo
+// see https://www.kernel.org/doc/Documentation/networking/proc_net_tcp.txt
+func readSocketInfo(procPIDPath string) map[uint64]socketInfo {
+	si := make(map[uint64]socketInfo)
+
+	fs, err := procfs.NewFS(procPIDPath)
+	if err != nil {
+		log.Debugf("Failed to read %s: %s", procPIDPath, err)
+		return si
+	}
+
+	for protocol, parser := range map[string]func() (procfs.NetTCP, error){
+		"tcp":  fs.NetTCP,
+		"tcp6": fs.NetTCP6,
+	} {
+		addrs, err := parser()
+		if err != nil {
+			log.Debugf("Failed to read %s socket info in %s: %s", protocol, procPIDPath, err)
+			continue
+		}
+		for _, entry := range addrs {
+			si[entry.Inode] = socketInfo{
+				fmt.Sprintf("%s:%d->%s:%d", entry.LocalAddr, entry.LocalPort, entry.RemAddr, entry.RemPort),
+				stateStr(entry.St),
+				protocol,
+			}
+		}
+	}
+
+	for protocol, parser := range map[string]func() (procfs.NetUDP, error){
+		"udp":  fs.NetUDP,
+		"udp6": fs.NetUDP6,
+	} {
+		addrs, err := parser()
+		if err != nil {
+			log.Debugf("Failed to read %s socket info in %s: %s", protocol, procPIDPath, err)
+			continue
+		}
+		for _, entry := range addrs {
+			si[entry.Inode] = socketInfo{
+				fmt.Sprintf("%s:%d->%s:%d", entry.LocalAddr, entry.LocalPort, entry.RemAddr, entry.RemPort),
+				stateStr(entry.St),
+				protocol,
+			}
+		}
+	}
+
+	unix, err := fs.NetUNIX()
+	if err == nil {
+		for _, entry := range unix.Rows {
+			si[entry.Inode] = socketInfo{
+				fmt.Sprintf("%s:%s", entry.Type, entry.Path),
+				fmt.Sprintf("%s:%s", entry.State, entry.Flags),
+				"unix",
+			}
+		}
+	} else {
+		log.Debugf("Failed to read unix socket info in %s: %s", procPIDPath, err)
+	}
+
+	return si
+}
+
+// fileStats returns the type, permission, size, and inode of a file
+func fileStats(statf func(string) (os.FileInfo, error), path string) (string, string, int64, uint64) {
+	stat, err := statf(path)
+	if err != nil {
+		log.Debugf("stat failed for %s: %v", path, err)
+		return "", "", 0, 0
+	}
+
+	ty := "REG"
+	if stat.Mode()&os.ModeSocket != 0 {
+		ty = "SOCKET"
+	} else if stat.Mode()&os.ModeNamedPipe != 0 {
+		ty = "PIPE"
+	} else if stat.Mode()&os.ModeDevice != 0 {
+		ty = "DEV"
+	} else if stat.Mode()&os.ModeDir != 0 {
+		ty = "DIR"
+	} else if stat.Mode()&os.ModeCharDevice != 0 {
+		ty = "CHAR"
+	} else if stat.Mode()&os.ModeSymlink != 0 {
+		ty = "LINK"
+	} else if stat.Mode()&os.ModeIrregular != 0 {
+		ty = "?"
+	}
+
+	size := stat.Size()
+	perm := stat.Mode().Perm().String()
+
+	var ino uint64
+	if sys, ok := stat.Sys().(*syscall.Stat_t); ok {
+		ino = sys.Ino
+	}
```

Comment:
Can you explain what is this code particularly doing?

Reply:
I added a comment to make it a little clearer, basically `Sys()` returns the real type returned by the syscall, this is OS dependent, so I have to make a type assertion to get that underlying type

And the reason I need the underlying type in the first place is because the inode number is not part of the exported interface (it's an OS-specific "implementation detail" let's say), so I have to get it from the os-specific type